### DISCUSSION TITLE
Passing JSON to artisan command  exposed via console layer

### DISCUSSION THREAD

**Discussion by vladifg** [brefphp/bref#1107] [Support]
Hello,

I am trying to pass JSON to my lambda that looks as following:

```
  helloWorld:
    handler: artisan_test
    environment:
      ARTISAN_COMMAND: my command
    layers:
      - ${bref:layer.php-80}
      - ${bref:layer.console}
```
      
My custom artisan_test handler

```
$kernel = $app->make(
    'Illuminate\Contracts\Console\Kernel'
);
print_r($_SERVER);
print_r($_SERVER['argv']);

$status = $kernel->handle(
    $input = new Symfony\Component\Console\Input\ArgvInput($command ? ['lambda', $command] : null),
    new Symfony\Component\Console\Output\ConsoleOutput
);

exit($status);
```

What I struggle here is that if I pass a string to my lambda, then $_SERVER['argv'] gets populated, but if I pass JSON, then $_SERVER is empty. 

I know that I can use the Bref Event pattern (https://bref.sh/docs/function/handlers.html), and this is what I do now, but it does not work for me since I can not return non-zero error codes. All my lambdas are treated as success regardless of the exit code.

Is there a way to achieve what I want apart of passing JSON as encoded string, and also being able to return an error code? It will be super nasty and will make my step functions way more complex.

Thank you!

**Comment by mnapoli** [comment]
> I know that I can use the Bref Event pattern (https://bref.sh/docs/function/handlers.html), and this is what I do now, but it does not work for me since I can not return non-zero error codes. All my lambdas are treated as success regardless of the exit code.

With a function/class handler, anything returned is a success, and any exception thrown is a failure. So you should be able to throw an exception and be done with it.

But anyway, regarding your example it looks incomplete: the `ARTISAN_COMMAND` environment variable is never used? And then in your handler you use `argv` which doesn't make sense here.

Do you want a function where the CLI arguments of the artisan command are hardcoded?

Or do you want to pass the arguments (the JSON thing) dynamically when invoking the function?

Also how do you invoke the function?

It seems to me you are trying to achieve something in a way that is overcomplicated, maybe if you explain what you are trying to achieve we can find a better solution?

**Reply by vladifg** [reply]
Hello, you are absolutely correct. This was me experimenting and not a final solution, so I copied over a broken thing. The below is what I have in use now, and it does not error when exception is thrown. The output of the lambda contains the exception, but it is still all green as if it succeeded.

This Is my serverless
```
myFunction:
    handler: lambda
    environment:
      ARTISAN_COMMAND: upload-dispatch-confirmation
    layers:
      - ${bref:layer.php-80} # PHP
```


And this is the file that is run

```#!/usr/bin/env php
<?php
use Bref\Context\Context;
use Illuminate\Console\BufferedConsoleOutput;
use Symfony\Component\Console\Input\ArgvInput;
use Symfony\Component\Console\Output\ConsoleOutput;

require __DIR__ . '/vendor/autoload.php';

/**
 * Lambda runner. This is here because the default artisan command was doing something
 * odd and undocumented when parsing json as an input. This allows to pass json
 * to our lambda commands
 */

return function ($event, Context $context) {
    $app = require __DIR__ . '/bootstrap/app.php';
    $kernel = $app->make(
        'Illuminate\Contracts\Console\Kernel'
    );


    $command = env('ARTISAN_COMMAND');
    $input = json_encode($event);
    $args = [
        "lambda",
        $command,
        $input
    ];

    $output = new BufferedConsoleOutput();
    $kernel->handle(new ArgvInput($args), $output);
    return $output->fetch();
};
```

Even when the exception is thrown in `    $kernel->handle(new ArgvInput($args), $output);`, it still makes lambda as successful. I guess this is because I am not using the console layer.

The reason I am not using console layer is because I can not pass JSON to command unless it is an escaped string which makes debugging and manual running is very time consuming with escaping/unescaping manually

**Reply by vladifg** [reply]
I am trying to have my lambdas set up in a way that I can give them unescaped JSON, i.e. `{"name":"Bref"}`, and so that all exceptions would result in the execution being marked as failed

**Reply by mnapoli** [reply]
I think there's a flag to enable so that the Console kernel throws an exception (when there's an error) instead of catching it. You might need to research how to enable that.
